### ISSUE TITLE
Bug Report: setup_gemdaq.sh kills ssh session

### ISSUE THREAD

**Issue by bdorney** [cms-gem-daq-project/sw_utils#3]
## Brief summary of issue

Occasionally sourcing `setup_gemdaq.sh` kills `ssh` session:

```
% source sw_utils/scripts/setup_gemdaq.sh -p $PWD/venv
ERROR: You must use "source sw_utils/scripts/setup_gemdaq.sh" to run this script.











Connection to gem904qc8daq closed.
``` 

Here the script did not realize it was being `source`'d and then it killed the `ssh` session.  A subsequent login and call succeeded:

```
% ssh -XtY <username>@gem904qc8daq
% source sw_utils/scripts/setup_gemdaq.sh -p $PWD/venv
source $PWD/venv/cc7/py2.7/bin/activate
(py2.7) %
```

### Types of issue

- [X] Bug report (report an issue with the code)
- [ ] Feature request (request for change which adds functionality)

## Expected Behavior

Should recogonize it's being `source`'d and if not it should not kill the `ssh` session.

## Current Behavior

Doesn't always recognize it's being `source`'d and then when it doesn't it kills the `ssh` session.

### Steps to Reproduce (for bugs)

1.  Login to a 904 DAQ machine
2. `source sw_utils/scripts/setup_gemdaq.sh -p $PWD/venv`

Appears to happen ~50% of the time.

## Possible Solution (for bugs)

Should use `return` here instead of `exit`:

https://github.com/cms-gem-daq-project/sw_utils/blob/07036f58adf2f1e2e7a49110c51ac035e4b0a18b/scripts/setup_gemdaq.sh#L4-L7

This will resolve the killing an `ssh` session.  However why it doesn't recognize it's being sourced is beyond me.

## Context (for feature requests)

Script should recognize it's being `source`'d

## Your Environment

* Version used: 07036f58adf2f1e2e7a49110c51ac035e4b0a18b
* Shell used: `/bin/zsh`

**Comment by lmoureaux** [comment]
Could a reproduce a few times without apparent reason, with both `bash` and `zsh` on `lxplus`. However didn't find a reliable way to trigger it.

In particular tried seven times:
```
$ ssh -tXY gem904qc8daq /usr/bin/zsh
% cd GEM
% export ELOG_PATH=$(readlink -f elog)
% source sw_utils/scripts/setup_gemdaq.sh
```
And five times:
```
$ ssh -tXY gem904qc8daq /usr/bin/zsh
% cd GEM
% export ELOG_PATH=$(readlink -f elog)
% source sw_utils/scripts/setup_gemdaq.sh -p $PWD/venv
```
Without the `ssh` session being killed by the script.

**Comment by bdorney** [comment]
Seems like a potentially isolated incident then.  However, change `exit 1` to `return 1` since an `exit` will kill the `SHELL`:

```
       exit [n]
              Cause the shell to exit with a status of n.  If n is omitted, the exit status is that of the last command executed.  A trap on EXIT is executed before the shell terminates.
```

While a `return`:

```
       return [n]
              Causes a function to exit with the return value specified by n.  If n is omitted, the return status is that of the last command executed in the function body.  If used outside a function, but during execution  of  a
              script  by  the .  (source) command, it causes the shell to stop executing that script and return either n or the exit status of the last command executed within the script as the exit status of the script.  If used
              outside a function and not during execution of a script by ., the return status is false.  Any command associated with the RETURN trap is executed before execution resumes after the function or script.
```

Will cause the execution to stop just returning a nonzero exit code (e.g. `return 1`).  Then I'll update `master`.

**Comment by lmoureaux** [comment]
> While a `return`: [...]

`return` won't work in case the script isn't sourced:
```
$ bash
$ return
bash: return: can only `return' from a function or sourced script
```

**Comment by bdorney** [comment]
And what about:

```
kill -INT $$
```

**Comment by lmoureaux** [comment]
Seems to work with `bash`

**Comment by bdorney** [comment]
@lmoureaux it also appears how you've ensured `source`'ing happens throws an error most of the time:

```
[gemuser@ctp7-pc DAQ]$ source sw_utils/scripts/setup_gemdaq.sh -h
Usage: source sw_utils/scripts/setup_gemdaq.sh [options]
    Options:
        -c cmsgemos release version (e.g. X.Y.Z)
        -d debug information is printed
        -g gemplotting release version (e.g. X.Y.Z)
        -G gemplotting dev version (e.g. single integer)
        -h displays this string
        -p Path to the venv location
        -P When at P5, port of a SOCKS proxy to be used by pip
        -v vfatqc release version (e.g. X.Y.Z)
        -V vfatqc dev version (e.g. single integer)
        -w No value following, deletes and recreates the venv from scratch

    The virtualenv found at -p will either be activated or created
[gemuser@ctp7-pc DAQ]$ source sw_utils/scripts/setup_gemdaq.sh -h
ERROR: You must use "source bash" to run this script.

If you used "source bash", please add a comment here:
    https://github.com/cms-gem-daq-project/sw_utils/issues/3
I generated some debugging information for you in /home/gemuser/DAQ/setup_gemdaq.debug
Please add it to your comment.

[gemuser@ctp7-pc DAQ]$ source sw_utils/scripts/setup_gemdaq.sh -h
ERROR: You must use "source bash" to run this script.

If you used "source bash", please add a comment here:
    https://github.com/cms-gem-daq-project/sw_utils/issues/3
I generated some debugging information for you in /home/gemuser/DAQ/setup_gemdaq.debug
Please add it to your comment.
```

**Comment by lmoureaux** [comment]
```
I generated some debugging information for you in /home/gemuser/DAQ/setup_gemdaq.debug
Please add it to your comment.
```

Could you please attach the file and/or provide steps to reproduce the bug?

**Comment by bdorney** [comment]
[setup_gemdaq.txt](https://github.com/cms-gem-daq-project/sw_utils/files/2127279/setup_gemdaq.txt)

You may also want to rename this to `setup_gemdaq_debug.txt` since `github` won't allow an upload of a `*.debug` file.

**Comment by lmoureaux** [comment]
The debug file helped me rule out issues in the environment or shell options.

I found this minimal example that shows the problem:
```
$ cat ./test
if ! $(return >/dev/null 2>&1); then echo "Not sourced"; fi
$ (exit 0)
$ source ./test
$ (exit 1)
$ source ./test
Not sourced
```
If someone understands what's going on here, please explain...

@bdorney Temporary workaround: try using:
```
(exit 0); source path/to/setup_gemdaq.sh
```
Please report any issue here; in the meantime I'm trying to find a proper fix.

**Comment by bdorney** [comment]
closed by #4
